File: Project/src/model/State.java
package model;

public class State{
    int f;
    int g;
    int h;
    State previous;
    boolean inOpen;
    int hashValue;


    void setF(){
        f = (int) (g + h);
    }
    void initNewState(int gValue){
        g = gValue;
        h = computeH();
        setF();
        previous = null;
        inOpen = false;
        hashValue = computeHash();
    }

    private int computeHash() {
        return 0;
    }

    private int computeH() {
        return 0;
    }

    int getF() {
        return f;
    }

    void setG(int gValue) {
        g = gValue;
        setF();
    }

    int getG() {
        return g;
    }

    int getH() {
        return h;
    }

    void setPrev(State s) {
        previous = s;
    }

    State getPrev() {
        return previous;
    }

    void addOpen() {
        inOpen = true;
    }

    void removeOpen() {
        inOpen = false;
    }

    boolean checkOpen() {
        return inOpen;
    }

     int getHash() {
        return hashValue;
    }

    boolean compareTo(State other) {
        if (getF() < other.getF()) return true;
        if ((getF() == other.getF()) && (getG() > other.getG())) return true;
        return false;
    }


}

File: FinalProject/FinalProject/state.cpp
#pragma once

#include <vector>
#include <string>
using namespace std;


class MyState {

public:
    int heapIdx;

    int getF() {
        return f;
    }

    void setG(int gValue) {
        g = gValue;
        setF();
    }

    int getG() {
        return g;
    }

    int getH() {
        return h;
    }

    void setPrev(MyState* s) {
        prev = s;
    }

    MyState* getPrev() {
        return prev;
    }

    void addOpen() {
        inOpen = true;
    }

    void removeOpen() {
        inOpen = false;
    }

    bool checkOpen() {
        return inOpen;
    }

    size_t getHash() {
        return hashval;
    }

    bool operator<(MyState& other) {
        if (getF() < other.getF()) return true;
        if ((getF() == other.getF()) && (getG() > other.getG())) return true;
        return false;
    }

    virtual vector<MyState*> getNeighbors() = 0;

    virtual bool operator==(const MyState& other) = 0;

    virtual string toString() = 0;

    virtual ~MyState() {}

    int f;
    int g;
    int h;
    MyState* prev;
    bool inOpen;
    size_t hashval;

    void initNew(int gval) {
        g = gval;
        h = computeH();
        setF();
        prev = nullptr;
        inOpen = false;
        hashval = computeHash();
    }

    void setF() {
        f = g + h;
    }

    virtual int computeH() = 0;

    virtual size_t computeHash() = 0;

};
